### PR TITLE
crypto/merkle: Add error handling

### LINKED ISSUE #557: Fix(merkle): broken error handling in ValueOp (VSA-2022-100)
Original issue [reported here](https://github.com/tendermint/tendermint/pull/10011). Thanks @lklimek for the PR.

> This is a security fix of https://blog.verichains.io/p/vsa-2022-100-tendermint-forging-membership-proof , backported from https://github.com/dashpay/tenderdash/pull/601
> It changes how errors are handled when generating proofs.

This is a non-issue for CometBFT itself and we are not aware of anyone using this particular library, but we are fixing it preemptively. Besides the changes by @lklimek, we are disallowing verification of `nil` roots; if this is a valid use case for anyone, then it must be handled outside the function.

### PR DESCRIPTION
**Pull request by lasarojc**
Closes #557 

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: crypto/merkle/proof.go [hunk 1]
```diff
@@ -60,15 +63,18 @@ func (sp *Proof) Verify(rootHash []byte, leaf []byte) error {
...
+	}
 	if !bytes.Equal(computedHash, rootHash) {
 		return fmt.Errorf("invalid root hash: wanted %X got %X", rootHash, computedHash)
 	}
 	return nil
 }
 
 // Compute the root hash given a leaf hash.  Does not verify the result.
-func (sp *Proof) ComputeRootHash() []byte {
+func (sp *Proof) ComputeRootHash() ([]byte, error) {
```
Comment (cason) [inline crypto/merkle/proof.go:77]: This method is not used by other packages in the CometBFT project. I wonder whether it should be made private.

Comment (lasarojc) [reply crypto/merkle/proof.go]: Thanks for the suggestion. Done!

File: .changelog/unreleased/breaking-changes/558-tm10011.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- Do not allow verifycation of Merkle Proofs against empty trees (`nil` root). `Proof.computeRootHash` is now private.
```
Comment (thanethomson) [inline .changelog/unreleased/breaking-changes/558-tm10011.md:1]:
```suggestion
- `[crypto/merkle]` Do not allow verification of Merkle Proofs against empty trees (`nil` root). `Proof.computeRootHash` is now private.
```

File: go.mod [hunk 3]
```diff
@@ -51,6 +51,7 @@ require (
 	github.com/gofrs/uuid v4.4.0+incompatible
 	github.com/google/uuid v1.3.0
 	github.com/oasisprotocol/curve25519-voi v0.0.0-20220708102147-0a8a51822cae
+	github.com/tendermint/go-amino v0.16.0
```
Comment (thanethomson) [inline go.mod:54]: Can we perhaps try to not import go-amino? I see it's just used in the new test case.

**Comment by thanethomson** [comment]
It seems we have to backport all the way back to v0.34, right?

**Comment by adizere** [comment]
> It seems we have to backport all the way back to v0.34, right?

Since this involves a Go API change, can it be safely backported without potentially breaking something in the code of users?

**Comment by lasarojc** [comment]
We can leave the renaming of the function to avoid the breaking on back ports.

**Comment by lasarojc** [comment]
@adizere I've wrapped the private function into a public one, that panics in the case of error. So there is a change of behavior, but not of interface.
This will be back ported.

In a future PR I will remove the public one.

**Review by thanethomson — APPROVED** [review]
Excellent, thanks @lasarojc!

File: .changelog/unreleased/breaking-changes/558-tm10011.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+- `[crypto/merkle]` Do not allow verification of Merkle Proofs against empty trees (`nil` root). `Proof.computeRootHash` is now private.
```
Comment (thanethomson) [inline .changelog/unreleased/breaking-changes/558-tm10011.md:1]:
```suggestion
- `[crypto/merkle]` Do not allow verification of Merkle Proofs against empty trees (`nil` root). `Proof.ComputeRootHash` is now panics when it encounters an error, but `Proof.Verify` does not panic
```

**Pull request merged**